vavr/src/main/java/io/vavr/collection/Array.java

```diff
@@ -984,21 +984,63 @@ public int length() {
 
     @Override
     public Array<Array<T>> permutations() {
-        if (isEmpty()) {
-            return empty();
-        } else if (delegate.length == 1) {
-            return of(this);
+        final Array<T> src = distinct();
```

Comment:
`EMPTY.distinct()` and `SINGLETON.distinct()` operations cost nothing. On other side, used algorithm requires length of source array at least 2, and we should check that result of `distinct()` operation is valid (to cover cases like [1,1,1,1...]). I would leave this as is.